src/lib/propagation/DAG.js

```diff
@@ -0,0 +1,63 @@
+import * as NodeFns from './nodeFns'
+import {SimulationNode} from './node'
+import * as constants from './constants'
+
+import * as _collections from 'gEngine/collections'
+
+export class SimulationDAG {
+  constructor(nodes) {
+    if (!!_.get(window, 'recorder')) { window.recorder.recordSimulationDAGConstructionStart(this) }
+    let rest = nodes.map(NodeFns.extractInputs)
+
+    let byId = _.transform(rest, (map, node) => {map[node.id] = {node, lastAncestors: node.inputs, ancestors: node.inputs}}, {})
```

Comment:
i'd rename ``map`` to ``resultMap``.  Change ``byId`` to ``remaining_nodes_by_id`` or something.  Don't mind being a bit more explicit, this code now is quite terse (though obviously well thought out).